lib/server.js

```diff
@@ -1,61 +1,40 @@
 'use strict'
 
-const assert = require('assert')
-const http = require('http')
-const https = require('https')
-
-const { kState, kOptions } = require('./symbols')
+const { kState, kOptions, kBindingServer } = require('./symbols')
+const { FST_ERR_CREATE_SERVER_MISSING_OPTIONS, FST_ERR_CREATE_SERVER_MISSING_HTTP_HANDLER } = require('./errors')
 const { FST_ERR_HTTP2_INVALID_VERSION, FST_ERR_REOPENED_CLOSE_SERVER, FST_ERR_REOPENED_SERVER } = require('./errors')
 
 function createServer (options, httpHandler) {
-  assert(options, 'Missing options')
-  assert(httpHandler, 'Missing http handler')
-
-  let server = null
-  if (options.serverFactory) {
-    server = options.serverFactory(httpHandler, options)
-  } else if (options.http2) {
-    if (options.https) {
-      server = http2().createSecureServer(options.https, httpHandler)
-    } else {
-      server = http2().createServer(httpHandler)
-    }
-    server.on('session', sessionTimeout(options.http2SessionTimeout))
-  } else {
-    // this is http1
-    if (options.https) {
-      server = https.createServer(options.https, httpHandler)
-    } else {
-      server = http.createServer(httpHandler)
-    }
-    server.keepAliveTimeout = options.keepAliveTimeout
-    server.requestTimeout = options.requestTimeout
-    // we treat zero as null
-    // and null is the default setting from nodejs
-    // so we do not pass the option to server
-    if (options.maxRequestsPerSocket > 0) {
-      server.maxRequestsPerSocket = options.maxRequestsPerSocket
-    }
+  if (options == null) {
+    throw new FST_ERR_CREATE_SERVER_MISSING_OPTIONS(typeof options)
   }
 
-  if (!options.serverFactory) {
-    server.setTimeout(options.connectionTimeout)
+  if (httpHandler == null) {
+    throw new FST_ERR_CREATE_SERVER_MISSING_HTTP_HANDLER(typeof httpHandler)
   }
 
+  const server = getServerInstance(httpHandler, options)
+
   return { server, listen }
 
   // `this` is the Fastify object
   function listen () {
     const normalizeListenArgs = (args) => {
+      const defaultHost = '127.0.0.1'
```

Comment:
IMHO this should be `'localhost'` instead

Reply:
For clarity, you're saying that the default option will inspect for a value of `'localhost'` and then bind to `'127.0.0.1'` and `'::1'` as a result of that inspection?

Reply:
Don't have a strong opinion on this to me looks a little bit more meaningful having a distinction between IPv4 and IPv6 for the default, this if we're referring the host to be used and not the one to be inspected, but I'm completely fine with that approach if everybody agrees :)

Reply:
I think the one that needs to be resolved is https://github.com/fastify/fastify/pull/3546/files#r774957324. Changing this to localhost will be the consequence of that change.